crates/ursa-rpc-service/src/api.rs

```diff
@@ -79,65 +91,188 @@ pub trait NetworkInterface: Sync + Send + 'static {
+    /// Get peers from the network
     async fn get_peers(&self) -> Result<HashSet<PeerId>>;
 
-    // get the addrresses that p2p node is listening on
+    /// Get the addresses that p2p node is listening on
     async fn get_listener_addresses(&self) -> Result<Vec<Multiaddr>>;
 }
-#[derive(Clone)]
+
+type PendingRequests = Arc<RwLock<HashMap<Cid, Vec<Sender<Result<()>>>>>>;
+
 pub struct NodeNetworkInterface<S>
 where
     S: Blockstore + Store + Send + Sync + 'static,
 {
     pub store: Arc<UrsaStore<S>>,
     pub network_send: Sender<NetworkCommand>,
     pub provider_send: Sender<ProviderCommand>,
+    pending_requests: PendingRequests,
+    client: Arc<Client>,
+    origin_config: OriginConfig,
+}
+
+impl<S> NodeNetworkInterface<S>
+where
+    S: Blockstore + Store + Send + Sync + 'static,
+{
+    pub fn new(
+        store: Arc<UrsaStore<S>>,
+        network_send: Sender<NetworkCommand>,
+        provider_send: Sender<ProviderCommand>,
+        origin_config: OriginConfig,
+    ) -> Self {
+        Self {
+            store,
+            network_send,
+            provider_send,
+            origin_config,
+            pending_requests: Arc::new(RwLock::new(HashMap::new())),
+            client: Arc::new(Client::new()),
+        }
+    }
+
+    /// Fetch content from the network
+    async fn get_network(&self, root_cid: Cid) -> Result<()> {
+        info!("Fetching cid {root_cid} from network");
+        let (send, recv) = oneshot::channel();
+
+        // Dispatch bloom filtered bitswap request
+        self.network_send.send(NetworkCommand::GetBitswap {
+            cid: root_cid,
+            sender: send,
+        })?;
+
+        recv.await?
+    }
+
+    /// Fetch content from the origin
+    async fn get_origin(&self, root_cid: Cid) -> Result<()> {
+        info!("Fetching cid {root_cid} from origin (ipfs)");
+        let pending = self.pending_requests.clone();
+        let (tx, mut rx) = unbounded_channel();
+        match self.pending_requests.write().await.entry(root_cid) {
+            Entry::Occupied(mut e) => {
+                // there is a concurrent request for this cid, just wait for the first one and return
+                e.get_mut().push(tx);
+                return rx
+                    .recv()
+                    .await
+                    .ok_or_else(|| anyhow!("Failed to receive status from channel"))?;
+            }
+            Entry::Vacant(e) => {
+                e.insert(vec![tx]);
+            }
+        }
+
+        // we are the first concurrent request for this cid
+        let client = self.client.clone();
+        let req = RequestBuilder::new(
+            Method::Get,
+            format!(
+                "https://{}/ipfs/{root_cid}",
+                self.origin_config.ipfs_gateway
+            )
+            .parse()?,
+        )
+        .header("Accept", "application/vnd.ipld.car")
+        .build();
+
+        let store = self.store.db.clone();
+        spawn(async move {
```

Comment:
```suggestion
        task::spawn(async move {
```

Reply:
Why the verbosity? there's only one spawn type used in the file

Reply:
Because the entire project uses that verbosity and it's not verbose similar to your comments above. This is a task spawn as opposed to something else that can be spawned.

Reply:
Sounds good just curious

Reply:
https://github.com/fleek-network/ursa/pull/278/commits/2863d7d5d1be45e473db00c728f6e6e6339b2fd2